### PR TITLE
[AERIE-1691] Return default arguments alongside effective argument failures

### PR DESCRIPTION
**Pull request by pcrosemurgy**
* **Tickets addressed:** AERIE-1691
* **Review:** By commit  
* **Merge strategy:** Merge (no squash)  

## Description
Adds default arguments alongside the `getModelEffectiveArguments `/`getActivityEffectiveArguments` failure response. This will allow the UI to implement default values within forms with the `getModelEffectiveArguments `/`getActivityEffectiveArguments` workflow.
See [AERIE-1691](https://jira.jpl.nasa.gov/browse/AERIE-1691) for more context.

## Verification

<details><summary>Expand</summary>
<p>

### Success Responses

With `examples/banananation`'s all optional arguments:
```gql
query TestGetModelEffectiveArguments {
  getModelEffectiveArguments(missionModelId: 1, modelArguments: {}) {
    success
    errors
    arguments
  }
}
```
→
```json
{
  "data": {
    "getModelEffectiveArguments": {
      "success": true,
      "arguments": {
        "initialDataPath": "/etc/os-release",
        "initialPlantCount": 200,
        "initialProducer": "Chiquita"
      }
    }
  }
}
```

### Failure Responses

With `BakeBananaBread`, where only the `temperature` parameter has a default value:
```gql
query TestGetActivityEffectiveArguments {
  getActivityEffectiveArguments(missionModelId: 1, activityTypeName: "BakeBananaBread", activityArguments: {}) {
    success
    errors
    arguments
  }
}
```
→
```json
{
  "data": {
    "getActivityEffectiveArguments": {
      "success": false,
      "errors": {
        "tbSugar": {
          "schema": {
            "type": "int"
          },
          "message": "Required argument for activity \"BakeBananaBread\" not provided: \"tbSugar\" of type ValueSchema.INT"
        },
        "glutenFree": {
          "schema": {
            "type": "boolean"
          },
          "message": "Required argument for activity \"BakeBananaBread\" not provided: \"glutenFree\" of type ValueSchema.BOOLEAN"
        }
      },
      "arguments": {
        "temperature": 350
      }
    }
  }
}
```

With `examples/config-without-defaults`'s all required arguments:
```gql
query TestGetModelEffectiveArguments {
  getModelEffectiveArguments(missionModelId: 1, modelArguments: {}) {
    success
    errors
    arguments
  }
}
```
→
```json
{
  "data": {
    "getModelEffectiveArguments": {
      "success": false,
      "errors": {
        "a": {
          "schema": {
            "type": "int"
          },
          "message": "Required argument for configuration \"Configuration\" not provided: \"a\" of type ValueSchema.INT"
        },
        "b": {
          "schema": {
            "type": "real"
          },
          "message": "Required argument for configuration \"Configuration\" not provided: \"b\" of type ValueSchema.REAL"
        },
        "c": {
          "schema": {
            "type": "string"
          },
          "message": "Required argument for configuration \"Configuration\" not provided: \"c\" of type ValueSchema.STRING"
        }
      },
      "arguments": {}
    }
  }
}
```

</p>
</details>

## Documentation
Updated Aerie GQL API documentation: https://github.com/NASA-AMMOS/aerie/wiki/Aerie-GraphQL-API-Software-Interface-Specification#query-for-activity-effective-arguments

## Future work
None.

### CODE REVIEW

**Review by camargo — APPROVED** [review]
Tested this locally and implemented a change on top of it in the UI. LGTM!

File: merlin-framework-processor/src/main/java/gov/nasa/jpl/aerie/merlin/processor/generator/MissionModelGenerator.java [hunk 1]
```diff
@@ -24,6 +24,7 @@
 import gov.nasa.jpl.aerie.merlin.protocol.model.ConfigurationType;
 import gov.nasa.jpl.aerie.merlin.protocol.model.MerlinPlugin;
 import gov.nasa.jpl.aerie.merlin.protocol.model.MissionModelFactory;
+import gov.nasa.jpl.aerie.merlin.protocol.types.MissingArgumentsException;
```
Comment (kroffo) [inline merlin-framework-processor/src/main/java/gov/nasa/jpl/aerie/merlin/processor/generator/MissionModelGenerator.java]: Looks like an accidental addition. This import isn't being used.

File: merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/http/MerlinBindings.java [hunk 2]
```diff
@@ -207,7 +207,7 @@ private void getModelEffectiveArguments(final Context ctx) {
       final var arguments = this.missionModelService.getModelEffectiveArguments(missionModelId, input.arguments());
 
       ctx.result(ResponseSerializers.serializeEffectiveArgumentMap(arguments).toString());
-    } catch (final MissingArgumentException ex) {
+    } catch (final MissingArgumentsException ex) {
       ctx.status(200)
```
Comment (kroffo) [inline merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/http/MerlinBindings.java:211]: I know this isn't part of your PR, but should this status really be 200 if it's serializing failures? There's more below too.

**Review by kroffo — APPROVED** [review]
Looks good to me overall. One minor issue to fix, and one comment to address, but nothing showstopping, so approved!

File: merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/http/MerlinBindings.java (hunk 2 above)
Comment (camargo) [reply merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/http/MerlinBindings.java:211]: Yes. If I recall correctly we discussed this a bit when this API was first implemented. The HTTP request is succeeding (200), but there are failures in the parameters being reported. See: https://github.jpl.nasa.gov/Aerie/aerie-gateway/pull/12#issuecomment-429000

**Review by mattdailis — APPROVED** [review]
Looks good to me! I was a bit skeptical of the exception with a builder pattern, but then I looked at the generated code, and it looked quite elegant. Good job.

File: merlin-framework-processor/src/main/java/gov/nasa/jpl/aerie/merlin/processor/generator/MissionModelGenerator.java (hunk 1 above)
Comment (kroffo) [reply merlin-framework-processor/src/main/java/gov/nasa/jpl/aerie/merlin/processor/generator/MissionModelGenerator.java]: If you fix this, feel free to resolve this comment yourself.

Comment (pcrosemurgy) [reply merlin-framework-processor/src/main/java/gov/nasa/jpl/aerie/merlin/processor/generator/MissionModelGenerator.java]: Good catch! Just fixed

**Comment by pcrosemurgy** [comment]
>I was a bit skeptical of the exception with a builder pattern, but then I looked at the generated code, and it looked quite elegant.

Good point, and I think it should get some skepticism.

I'm not too fond of using an exception as a required part of our UI <-> mission model workflow. The only alternative I could think of was to split `getEffectiveArguments` into 2 methods – and then complicate the already complex UI argument cobbling logic by requiring yet another endpoint to be hit.

I guess the argument for the pattern this PR uses is that we're not developing the mission model interface with UI concerns as the 1st priority, and that'll sometimes result in the UI relying on an exception to get ancillary data (such as returning default arguments when required args. are missing).

**Pull request merged**